spec/hotelsystem_spec.rb

```diff
@@ -0,0 +1,146 @@
+    it "returns an instance of reservation" do
+      expect(hotel_system.reserve_room(Date.today + 2, Date.today + 4, 3)).must_be_kind_of Hotel::Reservation
+    end
+
+    it "raises error when given invalid dates" do
+      expect {
+        hotel_system.reserve_room(Date.today, Date.today - 1, 3).must_raise ArgumentError
+      }
+    end
+
+    it "adds reservations to the @reservations array" do 
+      reservations_length = hotel_system.reservations.length
+      new_reservation = hotel_system.reserve_room(Date.today, Date.today + 3, 3)
+
+      expect(hotel_system.reservations.length - 1).must_equal reservations_length
+    end
+
+    it "raises an error with invalid dates" do 
+      expect {
+        hotel_system.reserve_room(Date.today, Date.today - 1, 3)
+      }.must_raise ArgumentError
+    end
+
+    it "raises error for unavailable room" do
+      expect {
+        hotel_system.reserve_room(Date.today, Date.today + 1, 1)
+      }.must_raise ArgumentError
+    end
+  end
+
+  describe "HotelSystem mananges reservations by date" do
+    let (:reservation_matches) {
+      hotel_system.reservations_by_date(Date.today + 1)
+    }
+
+    it "returns array of all reservations" do
+      reservation_matches.each do |reservation|
+        expect(reservation).must_be_kind_of Hotel::Reservation
+      end
+    end
+
+    it "returns the proper number of reservations" do
+      expect(reservation_matches.length).must_equal 2
+    end
+
+    it "returns empty array if there are no matche÷s" do
+      no_reservations = hotel_system.reservations_by_date(Date.today + 20)
+      expect(no_reservations).must_be_kind_of Array
+      expect(no_reservations.length).must_equal 0
+    end
+  end
+
+  describe "accurately manages occupied rooms" do
+    it "returns instance of an occupied room" do
+      expect(hotel_system.occupied_rooms_list(Date.today, Date.today + 2)[0]).must_equal first_reservation.room
+    end
+
+    it "returns empty array when not occupied" do
+      no_reservations = hotel_system.occupied_rooms_list(Date.today + 18, Date.today + 20)
+      expect(no_reservations).must_be_kind_of Array
+      expect(no_reservations.length).must_equal 0
+    end
+
+    it "returns array of unavailable room numbers" do 
+      expect(hotel_system.occupied_rooms_list(Date.today, Date.today + 3)).must_be_kind_of Array
+    end
+
+    # date ranges
+    it "returns correct list of unavailable rooms scenarios" do 
+      first_room = [1]
+      second_room = [2]
+      both_rooms = first_room + second_room
+
+      expect(hotel_system.occupied_rooms_list(Date.today + 1, Date.today + 7)).must_equal both_rooms
+      expect(hotel_system.occupied_rooms_list(Date.today + 2, Date.today + 3)).must_equal second_room
+      expect(hotel_system.occupied_rooms_list(Date.today - 3, Date.today + 1)).must_equal first_room
+      expect(hotel_system.occupied_rooms_list(Date.today + 2, Date.today + 7)).must_equal second_room
```

Comment:
It might be wise to break each of these out as a separate test. That way if one fails, you know exactly what went wrong and have a nice human-readable name for it.